### ISSUE TITLE
question... init player

### ISSUE THREAD

**Issue by smolleyes** [RSATom/WebChimera#76]
hi

i put the player in a hidden by default in my app like this

``` 
<div id="playerContainer" style="display:none;">
		<object id="Player" type="application/x-chimera-plugin" style="height:100%;width:100%;">
		<param name="debug" value="true" />
		<param name="qmlsrc" value="player/themes/sleek/main.qml" />
		<param name="target" value="" />
		</object>
	</div>
```

and in my js i have a simple 

> player = document.getElementById('Player');

until this div is not visible or focused it seems i can t use the js api on it, how can i do ?

thanks

**Comment by jaruba** [comment]
Read down from here: https://github.com/RSATom/WebChimera/issues/34#issuecomment-65058889

about @gyzerok 's method (not iframe shim), it should help you solve your problem.

**Comment by RSATom** [comment]
yes, it's possible some browsers don't create plugin if it's hidden. It needs some trick to workaround it. I'm already met something similar, just need to remember where exactly.

**Comment by RSATom** [comment]
Found it, look this discussion: https://code.google.com/p/fbvlc/issues/detail?id=61

**Comment by jaruba** [comment]
Also, you can make the plugin window have 1px height and 1px width. It would still be loaded, just practically invisible. :)

**Comment by smolleyes** [comment]
hello

ok just set the main div to visibility:hidden and it works not display:none :p

**Comment by smolleyes** [comment]
ok so it works, but if i change tab the stream completely stop and attachd events are not working, for exemple

> player.addEventListener('MediaPlayerStopped', function() {

do not works, it works if i declare it after player.play('.....')

**Comment by RSATom** [comment]
What browser do you use?

**Comment by smolleyes** [comment]
sorry, events are working i had some errors before my declaration of the events... 

i use node-webkit 0.9.2 now so  chromium: "32.0.1700.107"

**Comment by RSATom** [comment]
Is problem with tab change still  exists?

**Comment by smolleyes** [comment]
ok forgot my comment on player stooping when i change tab too :p

it works, finally i just have a problem with the total length of the file (but i have the same problem in vlc) ffmpeg can get the real duration of the file, no way to force the duration in the player ? (in seconds)

**Comment by smolleyes** [comment]
duration with ffmpeg

```
C:\Users\smo>ffprobe -i http://192.168.0.36:54566/  -show_format -v quiet
[FORMAT]
filename=http://192.168.0.36:54566/
nb_streams=2
nb_programs=0
format_name=avi
format_long_name=AVI (Audio Video Interleaved)
start_time=0.000000
duration=6505.656000
size=737477910
bit_rate=906875
probe_score=100
TAG:encoder=Lavf55.18.104
[/FORMAT]
```

don t understand how or if i can acces the wjs object i see in the qml file from js ?

**Comment by RSATom** [comment]
Could you try vlc player nightly build? Maybe it's fixed in recent versions.

**Comment by smolleyes** [comment]
ok i try, brb

(it works on linux in vlc with peerflix)

**Comment by RSATom** [comment]
Sorry, don't  know what wjs is

**Comment by smolleyes** [comment]
i try the msi from here

http://nightlies.videolan.org/build/win32/vlc-3.0.0-20150208-2058/

sorry too lol, i discover qml ... it s vlcPlayer is see it in the qml file, no way to acces it from js ?

**Comment by smolleyes** [comment]
humm just start a command line Oo

**Comment by RSATom** [comment]
https://github.com/RSATom/WebChimera/wiki/JavaScript-API
All this API related to vlcPlayer instance in qml. I.e. It's no matter on which side you call functions (it has some exceptions but you could think about it as described )

**Comment by RSATom** [comment]
Do you mean, is it possible run ffprobe from qml?

**Comment by smolleyes** [comment]
don t know, i call it thru nodejs so it s not a problem but don t know if i can tell vlc or qml tu use the duration i give to him

**Comment by RSATom** [comment]
Understand, could you please check if "position" property show correct value(it has 0 to 1 interval)?

**Comment by smolleyes** [comment]
player.time
103709
player.time
104549

that s why it change all the time, not way to not refresh it ? (think thru qml if i understand)

**Comment by smolleyes** [comment]
is see this in UIsettings.qml

Component.onCompleted: { ui = variables; done = true; }
	Timer {
         interval: 0; running: done === true && mutebut.width != ui.settings.toolbar.buttonMuteWidth ? true : false; repeat: false
         onTriggered: { ui = ui }
    }
	Timer {
         interval: 0; running: done === true && prevBut.width != buttonWidth ? true : false; repeat: false
         onTriggered: { buttonWidth = buttonWidth }
    }

??

**Comment by smolleyes** [comment]
where can i find the "position" you asked for ? :)

**Comment by RSATom** [comment]
This value I get from libvlc, so i can't change it.

**Comment by jaruba** [comment]
So much happened in the meanwhile..

To use wjs you need to include "player/webchimera.js" in your page, then if the plugin object has id "webchimera" for example, then you call it with wjs("#webchimera")

Where did you download WebChimera Player from? The download page or did you download a demo directly? (need to know in order to help you overwrite the total length of a video element)

**Comment by smolleyes** [comment]
i understand, but could we disable it and get it ourself ? (thru ffprobe for me) then sent is a time

**Comment by RSATom** [comment]
"position" = player.position

**Comment by smolleyes** [comment]
i downloaded the ratio demo jaruba

**Comment by smolleyes** [comment]
player.position
0.014723594300448895

it works

**Comment by jaruba** [comment]
ok, open the file "player\core\Functions.qml"

and look for these lines:	
	
	// Start Get Video Length in seconds
	function getLength() {
		if (vlcPlayer.length > 0) {
			return vlcPlayer.length;
		} else {
			return vlcPlayer.time * (1 / vlcPlayer.position);
		}
	}
	// End Get Video Length in seconds

Change it to:
	
	// Start Get Video Length in seconds
	function getLength() {
		if (vlcPlayer.length > 0) {
			return vlcPlayer.length;
		} else if (settings.customLength > 0) {
			return settings.customLength;
		} else {
			return vlcPlayer.time * (1 / vlcPlayer.position);
		}
	}
	// End Get Video Length in seconds

**Comment by RSATom** [comment]
I think @jaruba will help better, I have go to sleep. Will back around 10 hours

**Comment by smolleyes** [comment]
ok good night RSATom, thanks :)

**Comment by smolleyes** [comment]
ok done jaruba, and now how can i set it from js in the app ?

**Comment by jaruba** [comment]
Now go to "player\core\Settings.qml" and add:

    	property var customLength: 0;

where you see all the other "property var"..

Now we have a property ready to accept the custom value, all we need to do is send it from on page js to qml. So we go back to "player/core/Functions.qml" and look for:

		} else {
			if (startsWith(message,"[start-subtitle]")) subMenu.playSubtitles(message.replace("[start-subtitle]","")); // Get Subtitle URL and Play Subtitle
			if (startsWith(message,"[clear-subtitle]")) subMenu.clearSubtitles(); // Clear Loaded External Subtitle
			if (startsWith(message,"[load-m3u]")) playM3U(message.replace("[load-m3u]","")); // Load M3U Playlist URL

And we add one more line to it:

		} else {
			if (startsWith(message,"[start-subtitle]")) subMenu.playSubtitles(message.replace("[start-subtitle]","")); // Get Subtitle URL and Play Subtitle
			if (startsWith(message,"[clear-subtitle]")) subMenu.clearSubtitles(); // Clear Loaded External Subtitle
			if (startsWith(message,"[load-m3u]")) playM3U(message.replace("[load-m3u]","")); // Load M3U Playlist URL
			if (startsWith(message,"[set-total-length]")) settings.customLength = message.replace("[set-total-length]",""); // Set custom total length

**Comment by jaruba** [comment]
Now, if you use wjs, you can use:

    wjs("#Player").plugin.emitJsMessage("[set-total-length]"+totalLength);

or without wjs:

    document.getElementById('Player').emitJsMessage("[set-total-length]"+totalLength);;
    
These examples presume you have a plugin object with id "Player" and that you have the total length in milliseconds saved in a variable by the name of "totalLength".

**Comment by smolleyes** [comment]
ok i understand now i call .emitJsMessage( message ) ... that s it

**Comment by smolleyes** [comment]
yes i have a player object i ll try now :) THANKS

and it help me understand the qml files... :palm_tree:

**Comment by jaruba** [comment]
Also, you should hook to the event that is triggered when the video ends and use:

    wjs("#Player").plugin.emitJsMessage("[set-total-length]0");

or:

    document.getElementById('Player').emitJsMessage("[set-total-length]0");

Otherwise, when the video has ended, the next video will probably have the same total length (because it overwrites all other properties except the one from vlcPlayer itself).

**Comment by smolleyes** [comment]
player.emitJsMessage("[set-total-length]"+6505);

it gives me currentTime / 00:06 Oo

**Comment by jaruba** [comment]
yeah, because it's in milliseconds, so 1000 milliseconds = 1 second

**Comment by smolleyes** [comment]
oh ok i ll try

**Comment by jaruba** [comment]
Also, I recommend you use the recommended method ( http://wiki.webchimera.org/Player_Embedding ) of embedding WebChimera Player, it gives a lot more options ( http://wiki.webchimera.org/Player_JavaScript_API ) and accepts all the plugin javascript api functions too.

**Comment by smolleyes** [comment]
ok it works :+1: 

now i ll search to rehresh the progress bar, exemple i have 1:48:25 as total time but seeking in the bar stop at 43:25

**Comment by smolleyes** [comment]
ok i ll add the js file, but can still use the visibility:hidden like this ?

**Comment by jaruba** [comment]
The thing is, that the progress bar has nothing to do with .time or .length, it prints the value of .progress which has values ranging from 0.0 (0%) to 1.0 (100%) of the plugin area length.

If you want to overwrite this functionality, first you will need to make sure that you ALWAYS have the correct current time and total length available. Set a new property in "player/core/Settings.qml" (like before), then calculate the new progress by doing (do this in the "onTime()" function in "player/core/Functions.qml"):

    settings.newProgress = getLength() / vlcPlayer.time;

and change all "vlcPlayer.progress" to "settings.newProgress" everywhere.

**Comment by smolleyes** [comment]
ok thanks, i m adding the webchimera.js now

but have the QML file form lcocal machien warning... fixing it :p

**Comment by jaruba** [comment]
Yeah, you should be able to use '<div id="playerContainer" style="visibility:hidden;">' without problems, because the player container is exposed in html. (not dynamically created like the plugin object).

**Comment by jaruba** [comment]
To fix the local QML file warning, just remove:

	// if page on local machine, add warning
	var globalurlstring = " ";
	var localwarning = '<div id="warning-wrapper"><div id="lwarning" class="btn">QML File cannot be loaded from your Local Machine! Upload the Demo on a Web server to see it working correctly!</div></div>';
	var wjsScripts = document.getElementsByTagName("script"),
		webchimeraSrc = wjsScripts[wjsScripts.length-1].src,
		webchimeraFolder = webchimeraSrc.substring(0, webchimeraSrc.lastIndexOf("/"));
	switch(window.location.protocol) {
	   case 'http:': break;
	   case 'https:': break
	   case 'file:':
		 document.body.innerHTML += localwarning;
		 break;
	   default: 
		 document.body.innerHTML += localwarning;
	}
	// end if page on local machine, add warning

from "player/webchimera.js"

**Comment by smolleyes** [comment]
i have webchimeraFolder is not defined if i do that

i try to add  case 'app:': break; but not working

**Comment by jaruba** [comment]
I changed my comment above about how to force the new progress. It now includes a hack that ensures that the "settings" property will be notified in qml each time it changes.

Try removing only:

	switch(window.location.protocol) {
	   case 'http:': break;
	   case 'https:': break
	   case 'file:':
		 document.body.innerHTML += localwarning;
		 break;
	   default: 
		 document.body.innerHTML += localwarning;
	}

See if that helps. I have to run out for an hour or so, leave any questions you may have here and I'll answer them as soon as I get back. :)

**Comment by smolleyes** [comment]
ok jaruba thanks a lot !! really cool :)

i try to fix the qml loading for the moment :p

**Comment by jaruba** [comment]
Did removing those line help load webchimera.js correctly?

**Comment by jaruba** [comment]
I'm thinking that changing:

	// if page on local machine, add warning
	var globalurlstring = " ";
	var localwarning = '<div id="warning-wrapper"><div id="lwarning" class="btn">QML File cannot be loaded from your Local Machine! Upload the Demo on a Web server to see it working correctly!</div></div>';
	var wjsScripts = document.getElementsByTagName("script"),
		webchimeraSrc = wjsScripts[wjsScripts.length-1].src,
		webchimeraFolder = webchimeraSrc.substring(0, webchimeraSrc.lastIndexOf("/"));
	switch(window.location.protocol) {
	   case 'http:': break;
	   case 'https:': break
	   case 'file:':
		 document.body.innerHTML += localwarning;
		 break;
	   default: 
		 document.body.innerHTML += localwarning;
	}
	// end if page on local machine, add warning

to this:

	// if page on local machine, add warning
	var globalurlstring = " ";
	var localwarning = '<div id="warning-wrapper"><div id="lwarning" class="btn">QML File cannot be loaded from your Local Machine! Upload the Demo on a Web server to see it working correctly!</div></div>';
	var wjsScripts = document.getElementsByTagName("script"),
		webchimeraSrc = wjsScripts[wjsScripts.length-1].src,
		webchimeraFolder = webchimeraSrc.substring(0, webchimeraSrc.lastIndexOf("/"));
	switch(window.location.protocol) {
	   case 'http:': break;
	   case 'https:': break
	   case 'app:': break
	   case 'file:':
		 document.body.innerHTML += localwarning;
		 break;
	   default: 
		 document.body.innerHTML += localwarning;
	}
	// end if page on local machine, add warning

should also fix it with one line (should add exception for "app:" protocol)

**Comment by smolleyes** [comment]
i tried jaruba but no

i mixed it all, added object html and the js file so i have 2 player objects ... and have all the methods and qml ok

**Comment by smolleyes** [comment]
http://wiki.webchimera.org/Player_JavaScript_API this miss a stop method i think

**Comment by jaruba** [comment]
Why would you want to stop the player? :)

I once made a hotkey that "stopped" the player, but it wasn't to helpful, also think that it sent the player to a state that you couldn't do anything with it anymore.

If you want to stop the player, you could do ```wjs("#webchimera").plugin.stop();```

(or if you downloaded the player before the last update to it last night, it would be ```wjs("#webchimera").video.stop();``` )

**Comment by smolleyes** [comment]
mainPlayer.video.stop() is ok yeah, i need it to know when to stop my torrent downloading ...

i m not successful with the custom progress bar length i m BAAAAADDD :p

so i added

// custom progress
settings.newProgress = getLength() / vlcPlayer.time;
settings = settings;

at the end of onTime in the functions.qml file

added 

property var newProgress: 0;

in Settings.qml

and replace all  "vlcPlayer.progress" to "settings.newProgress"

**Comment by jaruba** [comment]
it would probably only need you to change all "vlcPlayer.progress" to "settings.newProgress" in "player\themes\sleek\components\ProgressBar.qml"

**Comment by smolleyes** [comment]
i don t have any vlcPlayer.progress in progressbar.qml jaruba

**Comment by jaruba** [comment]
just got home, gimme a sec to look over the code to see where changes should be made.

**Comment by jaruba** [comment]
Your right, change all "vlcPlayer.position" to "settings.newProgress". (sorry, was not home)

I'll check some more code, I think some other changes might also be required for this to work.

**Comment by smolleyes** [comment]
ok thanks i ll try

does all this changes will be included in next release or i ll have to add them if  i update ?

**Comment by smolleyes** [comment]
doesn t work, the cursor is at 100 % right and time printed is still equal to amount of downloaded data i think (like he doesn t use our  customLength) :p

**Comment by jaruba** [comment]
Yeah, there are a bit more problems then you think, but I'm sure we can fix it.

So, now let's go back to "player\core\Functions.qml". Go to these lines:

	// Start Progress Bar Seek Functionality
	function progressDrag(mouseX,mouseY) {
		settings.dragging = true;
		var newtime = (vlcPlayer.time * (1 / vlcPlayer.position)) * ((mouseX -4) / theview.width);
		if (newtime > 0) timeBubble.srctime = getTime(newtime);
	}
	function progressChanged(mouseX,mouseY) {
		var newtime = (vlcPlayer.time * (1 / vlcPlayer.position)) * ((mouseX -4) / theview.width);
		if (newtime > 0) timeBubble.srctime = getTime(newtime);
	}
	function progressReleased(mouseX,mouseY) {
		lastPos = (mouseX -4) / theview.width;
		if (vlcPlayer.state == 6) {
			vlcPlayer.playlist.currentItem = lastItem;
			vlcPlayer.playlist.play();
		}
		vlcPlayer.position = lastPos;
		settings.dragging = false;
	}
	// End Progress Bar Seek Functionality

and change them to:

	// Start Progress Bar Seek Functionality
	function progressDrag(mouseX,mouseY) {
		settings.dragging = true;
		var newtime = (vlcPlayer.time * (1 / settings.newProgress)) * ((mouseX -4) / theview.width);
		if (newtime > 0) timeBubble.srctime = getTime(newtime);
	}
	function progressChanged(mouseX,mouseY) {
		var newtime = (vlcPlayer.time * (1 / settings.newProgress)) * ((mouseX -4) / theview.width);
		if (newtime > 0) timeBubble.srctime = getTime(newtime);
	}
	function progressReleased(mouseX,mouseY) {
		lastPos = (mouseX -4) / theview.width;
		if (vlcPlayer.state == 6) {
			vlcPlayer.playlist.currentItem = lastItem;
			vlcPlayer.playlist.play();
		}
		vlcPlayer.time = (vlcPlayer.time * (1 / settings.newProgress)) * ((mouseX -4) / theview.width);
		settings.dragging = false;
	}
	// End Progress Bar Seek Functionality

What we're doing here is that we're changing all "vlcPlayer.position" to "settings.newProgress" in all functions related to the progress bar. But the progress bar has to change "vlcPlayer.position" when mouse click is released (to jump through playback). But we don't have a reliable "vlcPlayer.position" anymore, so we will use "vlcPlayer.time" to set the new position by milliseconds instead of percent.

**Comment by smolleyes** [comment]
i see :)

but same problem cursor is at 100%, always showing 00:00:00 on bar hover

**Comment by smolleyes** [comment]
no changes to do in functions.js ?

**Comment by jaruba** [comment]
functions.js is part of the older version, you got a mixed repo because you downloaded from the website demo generator during a transitional phase. All .js files are now .qml files, so functions.js is now Functions.qml.

It's very strange that the Time Bubble shows 00:00:00. could you change:

    function progressChanged(mouseX,mouseY) {
        var newtime = (vlcPlayer.time * (1 / settings.newProgress)) * ((mouseX -4) / theview.width);
        if (newtime > 0) timeBubble.srctime = getTime(newtime);
    }

to:

    function progressChanged(mouseX,mouseY) {
        setText("test: "+settings.newProgress);
        var newtime = (vlcPlayer.time * (1 / settings.newProgress)) * ((mouseX -4) / theview.width);
        if (newtime > 0) timeBubble.srctime = getTime(newtime);
    }

and then tell me what appears in the notification text in the corner when dragging the position pointer?

**Comment by smolleyes** [comment]
ok, it say test: and value decrease while playing

**Comment by jaruba** [comment]
I don't understand, be a bit more specific. it should print "test: " and settings.newProgress, which should be from 0.0 (which should mean 0%) to 1.0 (which should mean 100%).

The value should increase while playing. So I need to know exact number it's printing to understand what I did wrong. :)

**Comment by jaruba** [comment]
Oh, I know what I did wrong, at the very start of this discution, I said you should add:

    settings.newProgress = getLength() / vlcPlayer.time;

to Functions.qml, it should of been:

    settings.newProgress = vlcPlayer.time / getLength();

Try changing that.

**Comment by smolleyes** [comment]
all right m8 :)

everything ok now !

**Comment by jaruba** [comment]
:dart: Yeah, I'll add a function like ".setTotalLength(float)" in the Player JavaScript API in the next version.

**Comment by smolleyes** [comment]
all right !

i have to go or my wife will kill me lol

really impressed by this plugin and really nice team and support, sure i ll have more questions soon :) 

thanks a lot jaruba see ya :)

**Comment by jaruba** [comment]
no problem, bye

**Comment by jaruba** [comment]
I downloaded node-webkit and played around with it. Managed to fix node-webkit support in WebChimera Player: https://github.com/jaruba/WebChimeraPlayer/commit/5b32bac9761742f63d6ec02d40a422a94600a82f

Example Node-Webkit App running the "Default Skin" demo:
www.webchimera.org/webchimera-player-nw.zip

I'll also add ".setTotalLength()" and release a new version.

**Comment by smolleyes** [comment]
all right jaruba, thanks, will llok at your demo :)

**Comment by jaruba** [comment]
I need your help to test the next version.

Could you download:
www.webchimera.org/webchimera-player-nw-smolleyes.zip

Then implement it in your setup? You should use .addPlayer() to add your video source. Then test if ```wjs("#webchimera").setTotalLength()``` and ```wjs("#webchimera").stopPlayer()```  works.

**Comment by smolleyes** [comment]
Hello 
I LL test it later today :) 

Brb asap !

**Comment by smolleyes** [comment]
i reboot on windows an try :p

**Comment by smolleyes** [comment]
hello

just made another dir with my lastest code and lastest nw.js

first, plugins works fine with the wjs method, cool

i ll try the 2 functions now

**Comment by jaruba** [comment]
great, hope the functions work too :)

**Comment by smolleyes** [comment]
i m trying but don t understand 

 The reason I need you to test it is because wjs("#webchimera").setTotalLength() sets the total length only if wjs("#webchimera").plugin.length returns 0

**Comment by smolleyes** [comment]
hummm

you added media with addPlaylist, what s the best way to do ?

clear the playlist then add new media ?

**Comment by jaruba** [comment]
Your initial question was: "it works, finally i just have a problem with the total length of the file (but i have the same problem in vlc) ffmpeg can get the real duration of the file, no way to force the duration in the player ? (in seconds)"

This happens because of a bug in libvlc that never updates the total length of the video, this bug is known and has a partial solution implemented in WebChimera Player, that calculates the total length based on time and position.

The problem you described happens to only a few videos, where length is broken in libvlc and the partial solution doesn't work eighter. So what I mean is, that is libvlc can return length correctly (like in the demos on the website), the value set with "setTotalLength()" will be ignored. It will only work when libvlc doesn't send the correct length. (because you shouldn't need to correct length if it already works)

**Comment by smolleyes** [comment]
ok first test

when i open the file  (without calling the function) i have 00:00:24 / ty:AN:AN

think it s normal no ?

**Comment by smolleyes** [comment]
possible to call addPlaylist('mymedia') with custom title ? 

i ll test the functions now :p

**Comment by jaruba** [comment]
If you use .addPlaylist multiple times, it will add the videos to the end of the playlist.

If you want to clear the entire playlist before adding a new video, you should use:

    wjs("#webchimera").plugin.playlist.clear();

before calling addPlaylist again

**Comment by jaruba** [comment]
.addPlaylist() has a "title" parameter ( http://wiki.webchimera.org/Playlist_Title_Parameter ), but I've done some tests today and it doesn't seem to work with peerflix links. I think this is because it takes to long to establish a valid connection to the peerflix links, and until a connection is established the playlist item isn't included.

I've made a small hack for this though, but I should also find a way to fix it. I'll explain everything when we make sure that the functions work and I release the new version.

**Comment by smolleyes** [comment]
i already have the title no problems

**Comment by jaruba** [comment]
How? I'm curious. :))

**Comment by jaruba** [comment]
"when i open the file (without calling the function) i have 00:00:24 / ty:AN:AN"

No, that shouldn't happen.. are you sure your sending the an int and not a string or null (or maybe even 0) through "setTotalLength()"?

**Comment by smolleyes** [comment]
yup sorry since it s my lastest code the video was coming from ffmpeg live transcoding, i m fixing it to send direct http stream from peerflix now (sorry)

**Comment by smolleyes** [comment]
stopPlayer is working but do not reset the progress bar :p

**Comment by jaruba** [comment]
did you send "setTotalLength()" before using "stopPlayer()"? Does it work if you don't use "setTotalLength()" before stopping?

**Comment by smolleyes** [comment]
it s reseted when i change the media, ok now it start very well and have the duration changing all the time

i add setTotalLength() like this, i mean no value before stopPlayer?

**Comment by smolleyes** [comment]
ok the function works well :)

wjs('#webchimera').setTotalLength(6509*1000) give me 01:48:29 and progress bar ok too

**Comment by jaruba** [comment]
Great! I'll just fix .stopPlayer() to reset the progress bar too.

**Comment by smolleyes** [comment]
so i have to write a function now, if wjs('#webchimera').plugin.time change more than , don t know... 5 time then i ll call my ffprobe function and set the detected time with your function :p

**Comment by smolleyes** [comment]
or will call it by default don t know ...

**Comment by smolleyes** [comment]
nice jaruba :D :+1:

**Comment by smolleyes** [comment]
just realize that i don t see the video lol, decoding problem i think, strange

**Comment by jaruba** [comment]
nope, ```wjs('#webchimera').plugin.time``` is not the issue, it's actually the only thing that is working correctly.

```wjs.("#webchimera").plugin.length``` is the problem, if you want to limit ffmpeg to only when it's needed, check:

    if (wjs.("#webchimera").plugin.length == 0) {
        // your code
    }

**Comment by smolleyes** [comment]
exact

wjs("#webchimera").plugin.length
-> 0

**Comment by smolleyes** [comment]
i ll look at the catchEvent on playing state to do this, or maybe i can find the length in peerflix infos

**Comment by smolleyes** [comment]
i send all my torrent to flix but open them before with read-torrent

so i have an object like this

```
Objectannounce: Array[29]announceList: Array[29]created: Mon Feb 16 2015 20:10:31 GMT+0100 (Paris, Madrid)files: Array[1]info: ObjectinfoBuffer: NativeBufferinfoHash: "8fadd0248a40260085d7ba14607a559fc2d00a20"lastPieceLength: 3790length: 733220558name: "[ www.Cpasbien.pw ] Grimoire.d.Arkandias.2014.FRENCH.DVDRip.XviD-UTT.avi"pieceLength: 262144pieces: Array[2798]private: false
```

you can get the name

**Comment by smolleyes** [comment]
so just can get the file length not the time

info: 
Objectlength: 733220558

will have to use ffprobe

**Comment by jaruba** [comment]
Do other videos work? You can check DebugView ( https://technet.microsoft.com/en-us/library/bb896647.aspx ) to see the plugin error log. But if the video has no feed, then it is most probably a libvlc bug and we will have to wait for a new version.

"Playing" state is best, I think "Opening" phase can never have ".plugin.length" because the video isn't loaded yet. But you need to make sure you check if ".plugin.length" is 0 only once per video, because "Playing" is triggered also when you pause / play the video.

**Comment by smolleyes** [comment]
i ll try with normal vlc first

**Comment by smolleyes** [comment]
ok for playing state and plugin.length 0

**Comment by jaruba** [comment]
What do you send to WebChimera Player? The "http://" path to the streaming video?

**Comment by smolleyes** [comment]
yes

**Comment by smolleyes** [comment]
vlc must be installed on the system to get the plugin working?.

**Comment by jaruba** [comment]
No, libvlc is part of the plugin.

**Comment by smolleyes** [comment]
ok, movie not working in vlc player neither... :p

**Comment by jaruba** [comment]
Could be a broken file..

**Comment by jaruba** [comment]
In about 15 minutes or so, I'll release the next version with working ".stopPlayer()" and explain how to force the title for peerflix links with a trick.

**Comment by smolleyes** [comment]
yes others works fine... i add my ffprobe function and have another "problem" but more gui related i ll have to understand the qml :p

in my currnet version i have a classic player bar at the bottom and another at the top display on mouse hover with statistics on the download/upload and connected peers AND a X to hide my player, have to do this and add a stop button always displayed too

**Comment by jaruba** [comment]
Except the button, the rest are pretty easy to do. Button isn't to hard eighter (if you want to make an image button), a button with icon fonts is a bit more complex in implementation.

**Comment by smolleyes** [comment]
just need a simple classic stop button and a way to hide the player

look my player here when you have time..., you have a video... http://streamstudio.me

it will be a good training on qml :p

**Comment by smolleyes** [comment]
humm so i did

```
wjs("#webchimera").catchEvent('QmlMessage', handle);

function handle(event) {
    if (event == "[MediaPlayerPlaying]") {
    	console.log("PLAYER START PLAYING !!!!!!!!!!!!");
    	if (wjs("#webchimera").plugin.length == 0 && mediaDuration == 0) {
			getVideoLength(currentMedia.link);
		}
    }
}

```

never see the PLAYER START PLAYING, something wrong ?

**Comment by jaruba** [comment]
"QmlMessage" is an event that only catches messages sent from QML to JS.

What you want to do is catch the "MediaPlayerPlaying" event instead. So this should work:

	wjs("#webchimera").catchEvent('MediaPlayerPlaying', handlePlaying);
	
	function handlePlaying(event) {
        console.log("PLAYER START PLAYING !!!!!!!!!!!!");
        if (wjs("#webchimera").plugin.length == 0 && mediaDuration == 0) {
            getVideoLength(currentMedia.link);
        }
	}

**Comment by smolleyes** [comment]
ah ok sorry bad understanding of the doc :/ , i try

**Comment by jaruba** [comment]
I probably didn't explain it to well eighter, I use qml messages as a form of events in the player.

**Comment by smolleyes** [comment]
ok, all done my function works :)

i see in the qml that the title bar is visible only in fullcreen, i ll try to make it visible the same way but not only in full screen, then add a X button or image on top right to hide my player, hard ?

**Comment by smolleyes** [comment]
have a strange bug, after few seconds the player stop reading
don t know why, works right in vlc

**Comment by jaruba** [comment]
You can download the latest version of WebChimera Player from here:

http://www.webchimera.org/download

or 

https://github.com/jaruba/WebChimeraPlayer/releases/tag/v1.00

If you update to this version, you should be able to set titles to all your peerflix streams by doing:

    wjs("#webchimera").addPlaylist({ url: localHref+filename })

Where "localHref" is the http link (ex: "http://localhost:8888/") and filename is the name of the file that will be played (eg: "my.file.mp4"). This hack will work because the http stream works regardless of the file you add after it, and the filename will be used inside WebChimera Player to determine the title of the video.

**Comment by smolleyes** [comment]
ok i download and will try

**Comment by smolleyes** [comment]
why not adding  function to pass an object ? like the addWithOptions method i saw in the wiki ?

**Comment by jaruba** [comment]
you can pass objects with .plugin.emitJsMessage(), just use JSON.stringify() on it to make it a string, then JSON.parse() on the other side to convert to JSON. :)

**Comment by smolleyes** [comment]
ok will try after :p

**Comment by jaruba** [comment]
Look at the "onMessage" function in Functions.qml, that's where the JSON is parsed, I'm already using JSON Objects for "settings" and "skinning", should be visible in the function.

**Comment by smolleyes** [comment]
humm

i tried wjs("#webchimera").addPlaylist({ url: localHref+filename })

i have the name it the playlist but it won t start reading

**Comment by jaruba** [comment]
do ```console.log("test: "+localHref+filename);```

**Comment by smolleyes** [comment]
test: http://192.168.0.36:54765/[ www.Cpasbien.pw ] 22. Minutes.2014.TRUEFRENCH.SUBFORCED.DVDRiP.XViD-ARTEFAC.avi

**Comment by jaruba** [comment]
I think it's because the filename has the space character in it, try:

    wjs("#webchimera").addPlaylist({ url: localHref+filename.split(" ").join("%20") })

**Comment by smolleyes** [comment]
%20 are parsed in the playlist ? 

(not working)

**Comment by jaruba** [comment]
By the way, about the Title Bar, you don't need to change anything inside the player to make it visible when it isn't in fullscreen. Check out: http://wiki.webchimera.org/TitleBar_Parameter

**Comment by smolleyes** [comment]
cool :)  i add it

**Comment by smolleyes** [comment]
nop won t play with the title

**Comment by jaruba** [comment]
this is so strange, try loading this hash: F706867CE38339A8A9ABC403AE886334C5A8E9DD

**Comment by smolleyes** [comment]
? :p

**Comment by jaruba** [comment]
updated comment

**Comment by smolleyes** [comment]
ok i resume

first, if i use the catchEvent, it works but once the length is changed the progress bar is not updated now AND the player stop reading few seconds later, totally frreezed

i just removed the catchEvent and now iy play normally

and now with just the link in addPlaylist, in the titlebar i have http://192 168 0 

that s all :/ strange

**Comment by jaruba** [comment]
Yeah, that's in the titleBar because that is the closest the player could find to a name for what it's playing. Curious problems indeed, you need to install debugView to see if there are any qml errors in the log that will point us in the right direction.

**Comment by smolleyes** [comment]
ok i search debugView

**Comment by jaruba** [comment]
I'm thinking it could be because the playing event is triggered a few times, not once, and ffmpeg might need some resources to do it's job, if it's called a few times in just a few milliseconds, it might freeze your app. You should add some sort of protection to ensure it's only called once per video element.

**Comment by smolleyes** [comment]
possible, i ll try brb

**Comment by smolleyes** [comment]
damn, very strange 

when i call my function the player freeze :/

**Comment by smolleyes** [comment]
how can i install debugView ?

**Comment by jaruba** [comment]
I found out why the title worked for me with that method and your's didn't. When I thought I was inserting "dummy text" in the link, that file actually existed on the local server. It seems peerflix exposes the files on the server. The reason it didn't work for you is that the file you tried does not exist on the server. This would be fixed if you could give the url to the exact file path to .addPlaylist().

Download from here:
http://download.sysinternals.com/files/DebugView.zip

Then run.

**Comment by smolleyes** [comment]
i have manyyyyyyyyy

[4224] file:///D:%5Cprojets%5Cstreamstudio%5Cnw12%5Cwin%5Ctest/player/core/Functions.qml:214: JSON.parse: Parse error

**Comment by jaruba** [comment]
what do you have on line 214 in Functions.qml?

is it:

	if (vlcPlayer.playlist.items[vlcPlayer.playlist.currentItem].setting) itemSettings = JSON.parse(vlcPlayer.playlist.items[vlcPlayer.playlist.currentItem].setting);

?

**Comment by smolleyes** [comment]
yup

**Comment by smolleyes** [comment]
no infos when the player freeze :/

**Comment by jaruba** [comment]
Change it to:

    if (isJson(vlcPlayer.playlist.items[vlcPlayer.playlist.currentItem].setting)) itemSettings = JSON.parse(vlcPlayer.playlist.items[vlcPlayer.playlist.currentItem].setting);

**Comment by jaruba** [comment]
Add a ```console.log("test");``` in the catchEvent() function, I'm curious how many times it's triggered during the freeze.

**Comment by smolleyes** [comment]
ok :)

**Comment by smolleyes** [comment]
i already have

function handlePlaying() {
    console.log("PLAYER START PLAYING !!!!!!!!!!!!");

and i see the message only once, so ok no ?

**Comment by jaruba** [comment]
Yeah, that's ok, then what could freeze the player? maybe ffmpeg freezes? the player doesn't have any reason to freeze.

**Comment by smolleyes** [comment]
i think he doesn t like process.spawn call, if i replace with process.exec it do not freeze but i can t get my infos with exec

**Comment by smolleyes** [comment]
maybe a nwjs bug (they use iojs now not node.js) i ll try older version to see

no more json errors :)

**Comment by jaruba** [comment]
how does it spawn the process? is it by any chance a cmd.exe process?

**Comment by smolleyes** [comment]
possible yes don t know the "undergound" of this functions

**Comment by jaruba** [comment]
I'm asking because it sounds like process.spawn is waiting for the echo response, while process.exec is doing it process asynchronously. If so, the the app doesn't freeze, it's just waiting for a response and the script can't continue until it gets it.

**Comment by smolleyes** [comment]
yes with exec not spawn it s non blocking 

heres my function 

```

function getVideoLength(link) {
  if(!upnpToggleOn) {
    link = decodeURIComponent(link);
  }
  
  if (process.platform === 'win32') {
    p = spawn(execDir+'/ffprobe.exe',[''+decodeURIComponent(link)+'', '-show_format','-v', 'quiet']); 
  } else {
    p = spawn(execDir+'/ffprobe',[''+decodeURIComponent(link)+'', '-show_format','-v', 'quiet']);
  }
  p.stdout.on('data',function(data){
    if(data.toString().indexOf('duration=') !== -1) {
      var rep = data.toString().match(/duration=(.*)/)[1];
      if(rep.indexOf('N/A') !== -1) {
        mediaDuration = 0;
      } else {
        mediaDuration = parseFloat(rep*1000);
      }
    }
  }); 
    p.on('exit',function(code){
    if(code === 0 && mediaDuration !== 0) {
        console.log('SET mediaLength to '+mediaDuration)
        wjs('#webchimera').setTotalLength(mediaDuration);
    }
  });
}

```

**Comment by smolleyes** [comment]
damn i can t get plugins from plugins dir with node-webkit 0.11.6 of before only nw12

**Comment by jaruba** [comment]
Well, the obvious (but messy) solution to this, would be to echo the response into a file (in cmd.exe it should be something like " > filename.txt" at the end of the call, it should be possible somehow with process.exec too) then read the new file after about a 7 second setTimeout() to get the echo response.

**Comment by smolleyes** [comment]
ok for plugins, same problem with 0.11.6 very very strange

**Comment by jaruba** [comment]
In debugView? does the video play correctly?

**Comment by smolleyes** [comment]
don t know nothing happen

[4296] [05:07:07.00484-  5036] raptr.exe       api.cpp(150)          OnProcessCreated               Process has been created: [32-bit, PID: 5608] D:\projets\streamstudio\nw12\win\test\nw.exe ("D:\projets\streamstudio\nw12\win\test\nw.exe" --type=plugin --plugin-path="C:\Users\smo\AppData\Roaming\WebChimera\0.2.3\npWebChimera.dll" --no-sandbox --lang=fr-FR --channel="3916.2.2051664585\1373723829" /prefetch:-390060480)
[5608] QImage::pixel: coordinate (0,1) out of range
[5608] QImage::pixel: coordinate (1,0) out of range
[5608] avi demux error: no key frame set for track 0
[5608] avi demux error: no key frame set for track 1
[5608] avi demux error: no key frame set for track 0
[5608] avi demux error: no key frame set for track 1
[5608] main vout display error: Failed to change zoom
[5608] main vout display error: Failed to set on top
[5608] main vout display error: Failed to change source AR
[5608] main vout display error: Failed to change zoom
[5608] main vout display error: Failed to change source AR

**Comment by jaruba** [comment]
Think RSATom could help you more with this.

**Comment by smolleyes** [comment]
damn

everything woks in node-webkit 0.9.2 but video is ugly (no acceleration i think )

**Comment by smolleyes** [comment]
so, all the bugs come from node-webkit not the player/plugin/my code .. :/

**Comment by jaruba** [comment]
You should use the latest version. This should be the solution to your problem with freezing:

http://stackoverflow.com/questions/4443597/node-js-execute-system-command-synchronously

From the post:

**UPDATE 2015-02-10**

AT LAST! NodeJS 0.12 supports execSync natively.
See official [docs]( http://nodejs.org/docs/v0.12.0/api/child_process.html#child_process_child_process_execsync_command_options )

**Comment by smolleyes** [comment]
i can t choose my nodejs with node-webkit :/ and they switched to io.js

**Comment by smolleyes** [comment]
https://iojs.org/api/child_process.html

**Comment by jaruba** [comment]
From the docs:

    var exec = require('child_process').exec,
    child;

    child = exec('cat *.js bad_file | wc -l',
      function (error, stdout, stderr) {
        console.log('stdout: ' + stdout);
        console.log('stderr: ' + stderr);
        if (error !== null) {
          console.log('exec error: ' + error);
        }
    });

```stdout``` is the output of the execution. So you can parse the data with that.

**Comment by smolleyes** [comment]
that s what i m trying yeah

**Comment by smolleyes** [comment]
same problem

the command works since it return 

 stdout: [FORMAT]
filename=http://192.168.0.36:54520/
nb_streams=2
nb_programs=0
format_name=avi
format_long_name=AVI (Audio Video Interleaved)
start_time=0.000000
duration=4748.000000
size=732686744
bit_rate=1234518
probe_score=100
TAG:encoder=MEncoder SVN-r33883(20110719-gcc4.5.2)
[/FORMAT]

flix.js:357 stderr: 

and i ear the video restarting in loop Oo

**Comment by jaruba** [comment]
Maybe it's because WebChimera and ffmpeg are both accesing the same file at the same time?

**Comment by smolleyes** [comment]
it s an http server so i don t think it could be a problem, i search :) damn

**Comment by jaruba** [comment]
Could be some conflict with ffmpeg and peerflix or ffmpeg and webchimera. But something could go wrong on the way. Did you hear the video in a loop before when it froze? Or was it acting differently then now?

**Comment by smolleyes** [comment]
no loop before exec command start it play normally

**Comment by smolleyes** [comment]
i ll try another way :p

**Comment by jaruba** [comment]
Not sure with how much more I can help with this problem. Don't think the player or the plugin have any errors here. Sounds like there's something wrong with ffmpeg or the method used to access it's data. Just needs some testing.

**Comment by jaruba** [comment]
Why don't you try using ffmpeg for node.js in your app? https://github.com/damianociarla/node-ffmpeg

It would avoid using cmd.exe for this all together.

**Comment by RSATom** [comment]
Guys, it seems you are worked all night... I don't understand even half of your discussion :)

**Comment by smolleyes** [comment]
raaaaaaaaaaaahhhh

yeah rsaTom hard work :p 

now when peerflix is ready i first get the length, then as callback i start playing, it works but even with timeout it won t set the length correctly

i save the length as timeDuration and in mediaplaying event callback i have

console.log("PLAYER START PLAYING !!!!!!!!!!!!",wjs("#webchimera").plugin.length, mediaDuration, videoLengthChecked );
    if (wjs("#webchimera").plugin.length == 0 && mediaDuration == 0 && !videoLengthChecked) {
    	videoLengthChecked = true;
    	setTimeout(function() {
		 wjs('#webchimera').setTotalLength(mediaDuration);
		},5000);
	}

in the console i have 
i have PLAYER START PLAYING !!!!!!!!!!!! 0 4748000 false

but it won t set the duration 

 wjs('#webchimera').setTotalLength(mediaDuration); works if i copy it by hand in the console, crazy

**Comment by smolleyes** [comment]
(seeking seems not working once .setTotalLength called and working by hand) i click the middle of the progress bar for exemple i just restart few seconds after where i am

**Comment by jaruba** [comment]
we're gonna have to start this up again tomorrow, falling asleep. :)

**Comment by smolleyes** [comment]
same thing :) good night m8, thanks !

**Comment by jaruba** [comment]
The code you made can't work, there are 2 issues with it:

- If you check ".plugin.length" before the video is opened, it will always be 0, even when it works. (because the media information has not been processed yet)

- When the video opens, it removes all changes made by ```.setTotalLength()```, this happens to all videos when they are opened (so the progress bar can start from 0 and not be influenced by the previous video's ```.setTotalLength()``` )

**Comment by smolleyes** [comment]
Hi jaruba 

This is exactly what i tried before we stopped :) 

Everything but the setTotalLength, if i do it by hand it works with mediaDuration saved as global, crazy

**Comment by jaruba** [comment]
I'll have to test this myself to understand the problem better, gimme a few days.

**Comment by jaruba** [comment]
Did you always test on a video that was already downloaded?

Because I've tested this thoroughly and I don't think it is possible. Not how you imagined it at least.

I've checked the speed of ffprobe, it usually runs between 90 to 400 milliseconds (and I'm on a laptop). If I use a lot of resources it can go up to 1000 milliseconds (1 second) maximum. There are very rare exceptions.

But ffprobe does not get the correct duration. I made the player check for the error and request a the new length, and ".format.duration" from ffprobe is printing a larger number every time it's called (on a 5 second setTimeout).

It managed to get the correct duration only after downloading about 60% of the video. So in this scenario, imagine that you have to recheck ffprobe constantly for (depending on your download speed and the size of the file) maybe around 6-10 minutes or so in order to get the correct duration.

I'm not sure if using ffprobe during playback had any impact on the video playback, I did have some issues but I think it may have been because the buffer parameter in .addPlayer() needed to have a higher value. I don't think this is a good solution for fixing this problem though, ffprobe has 30mb and checking every 30 - 60 seconds would still probably have an impact on resources.

I've also checked for alternatives to ffprobe and found libav, but libav is a lot bigger (160mb) and I don't think I can strip off to much excess to make it small enough for it to be worth trying.

**Comment by RSATom** [comment]
Some video formats does not store self length internally (usually it's streaming formats) - so sometimes it's not possible get correct video duration by design, until you don't download full media.
Can it be your case?

**Comment by jaruba** [comment]
This problem only happens with ".avi" files. And "vlcPlayer.length" is always 0 during the entire video.

**Comment by RSATom** [comment]
does it show correct 'position'?

**Comment by jaruba** [comment]
Yes and no..

This is the same issue I reported here: https://github.com/RSATom/WebChimera/issues/12

It happens to a lot of video sources to have "vlcPlayer.length" at 0 through the whole video, in response to this, WebChimera Player calculates the correct length by doing:

    vlcPlayer.time * (1 / vlcPlayer.position)

This usually works and gets the correct length. But only for ".avi" files it seems it doesn't. This is probably because "vlcPlayer.position" is not exact in these cases. (because vlcPlayer.time works as expected) So length just keeps increasing and changing constantly.

**Comment by RSATom** [comment]
.avi files contains such thing as "index", and it could be used only when full file available. If full file is not available players usually restore that "index" "on the fly", by result player could display duration of only downloaded part.
But earlier days (about 10 years ago) it was much worse - it was not possible play partially downloaded .avi file at all. Those days was special program to restore index of partially downloaded files to solve this problem.

**Comment by jaruba** [comment]
Then there will never be a fix for this in libvlc. Half a year has already past since I first found it and this has not been the first time this was mentioned to me. ffprobe seems like the only possible solution atm, at least it gets the real duration, even at 60% downloaded. I'll play around with it more, see if I can get this working in the safest way.

**Comment by RSATom** [comment]
It could be useful as I could be wrong, or something was changed last  years...

**Comment by smolleyes** [comment]
hello

i m ok with the index problem of avi files reported thru vlc but i just can say that any avi files played with popcorn time or my current app or sended thru dlna to my set top box (in streaming) always directly have the good length even for avi files so don t know... :/

**Comment by RSATom** [comment]
Then it's libvlc issue, and could be fixed only there. Or if I will switch to another multimedia engine...

**Comment by smolleyes** [comment]
jaruba seems having it working by using vlc 3 no ?

**Comment by smolleyes** [comment]
(good idear for the jitter chat :p)

**Comment by RSATom** [comment]
yes, you could use Vlc 3.0 but it in early development state, so you could meet many other issues...

**Comment by jaruba** [comment]
the index pointer works in vlc 3.0, the duration still doesn't work, but it's a lot more accurate, while now the duration never stabilises, in vlc 3.0 it gets to a duration that is very close to the real value pretty fast. Then the differences to the real value are about 1 second to 1 minute off. With some JS logic, I can make the bug look a bit better in vlc 3.0.

The solution I'm thinking of for this is not the best, but it should work. If we also use ffprobe (which as I stated, does not get the correct duration from the start), I can make it find the real duration in the first 2-5 minutes since video starts. When it gets the real duration, it should be sent to a server and inserted into a database along with hash and file name.

Then the next time the video is entered, just check back to the database, get the real duration and send it to WebChimera Player (no point for 50 users to use ffprobe when one has already done the job for all). This would fix the issue from the start of the video, without any need to increase resource use by all users.

**Comment by smolleyes** [comment]
hello jaruba

it seems you re working with node-webkit now lol :)

i wait for the linux version of webchimera to work on it

**Comment by jaruba** [comment]
Started as trying to understand everyone's problems in nw.js but evolved into an app. I should be done with it very soon.

Did you ever manage to fix the total duration issue? Do you need me to explain or try to make a demo of how to fix this best (as I explained in the previous comment)? This method requires you to have an online http server and make a simple api to save the data in a database.

**Comment by smolleyes** [comment]
I understood everything you said and i ll verify but in my current app i don t see wrong returns from ffprobe :/ strange 

Not a big fan of this api system but it s a solution

**Comment by smolleyes** [comment]
i just tried on few files, and removed ffprobe, just analysing ffmpeg output to catch the duration string

> total_data.toString().match(/Duration:\s\d\d:\d\d:\d\d\.\d\d/)

on the ffmpeg.stderr and it always have the exact good length :/

**Comment by jaruba** [comment]
Is your file already downloaded or are you downloading it from the beginning when you are using ffprobe? I can't get the correct duration at the start of downloading, it needs to get a bigger chunk of the file to get the right duration. How big is the file your trying this on, try on bigger files too (2-3 gb). Also try a few more avi files, some may react differently.

I've thought of a second method of fixing (kind of) the total duration bug.. By using vlc nightly to fix the progress seek bug, the total duration still changes constantly, but it is more precise, if I create a rule from which it only changes if there's a difference of at least 1 minute, it should get a pretty precise total duration (not exact though, but at a maximum of 1 minute difference). This is just a theory for now though. :)

**Comment by smolleyes** [comment]
yes some files start directly and some others are starting after 2-3% of downloading but i don t have anything to do just wait and it have the good length, i think ffmpeg "know" when he can start playing and have all informations needed...

**Comment by smolleyes** [comment]
1 min is not dramatic yeah :p

**Comment by jaruba** [comment]
This second solution is easy to test and tinker with until it works best for your case.

- Apply the [avi .time solution](https://github.com/RSATom/WebChimera/issues/80)

- Add the line ```property var memLastLength: 0;``` in ```player\core\Functions.qml```

- Also in ```player\core\Functions.qml``` change:

	// Start on State Changed
	function onState() {
		if (vlcPlayer.state == 1) {
			buftext.changeText = "Opening";

with:

	// Start on State Changed
	function onState() {
		if (vlcPlayer.state == 1) {
			buftext.changeText = "Opening";
			memLastLength = 0;

- In the same file (```player\core\Functions.qml```) also change this:

	// Start Get Video Length in seconds
	function getLength() {
		if (vlcPlayer.length > 0) {
			return vlcPlayer.length;
		} else if (settings.customLength > 0) {
			return settings.customLength;
		} else {
			return vlcPlayer.time * (1 / vlcPlayer.position);
		}
	}
	// End Get Video Length in seconds

with:

	// Start Get Video Length in seconds
	function getLength() {
		if (vlcPlayer.length > 0) {
			return vlcPlayer.length;
		} else if (settings.customLength > 0) {
			return settings.customLength;
		} else {
			var tempLength = vlcPlayer.time * (1 / vlcPlayer.position);
			if (memLastLength -60000 > tempLength || memLastLength +60000 < tempLength) {
				memLastLength = tempLength;
				return tempLength;
			} else return memLastLength;
		}
	}
	// End Get Video Length in seconds

It should be that simple. If you want to change the 1 minute difference to more / less, just change ```60000``` to a different millisecond value (in the changes made to the ```getLength()``` function). :)

**Comment by RSATom** [comment]
Can we close this issue?

**Comment by smolleyes** [comment]
Hello

Yes sure :)

**Comment by RSATom** [comment]
Ok, thanks.
